rust/client/src/ws/mod.rs

```diff
@@ -8,43 +9,46 @@ use tokio::sync::mpsc::Sender;
 use tokio_tungstenite::tungstenite::protocol::Message;
 use tokio_tungstenite::{connect_async, tungstenite::Utf8Bytes};
 
-use crate::{BACKPACK_WS_URL, BpxClient, DEFAULT_WINDOW, now_millis};
+use crate::{BpxClient, DEFAULT_WINDOW, Error, now_millis};
 
 impl BpxClient {
     /// Subscribes to a private WebSocket stream and sends messages of type `T` through a transmitter channel.
-    pub async fn subscribe<T>(&self, stream: &str, tx: Sender<T>)
+    pub async fn subscribe<T>(&self, stream: &str, tx: Sender<T>) -> Result<()>
     where
         T: DeserializeOwned + Send + 'static,
     {
         self.internal_subscribe(&[stream], tx).await
     }
 
     /// Subscribes to multiple private WebSocket streams and sends messages of type `T` through a transmitter channel.
-    pub async fn subscribe_multiple<T>(&self, stream: &[&str], tx: Sender<T>)
+    pub async fn subscribe_multiple<T>(&self, stream: &[&str], tx: Sender<T>) -> Result<()>
     where
         T: DeserializeOwned + Send + 'static,
     {
         self.internal_subscribe(stream, tx).await
     }
 
-    async fn internal_subscribe<T>(&self, stream: &[&str], tx: Sender<T>)
+    async fn internal_subscribe<T>(&self, stream: &[&str], tx: Sender<T>) -> Result<()>
     where
         T: DeserializeOwned + Send + 'static,
     {
         let timestamp = now_millis();
         let window = DEFAULT_WINDOW;
         let message = format!("instruction=subscribe&timestamp={timestamp}&window={window}");
+        let Some(signing_key) = &self.signing_key else {
+            return Err(Error::NotAuthenticated);
+        };
 
-        let verifying_key = STANDARD.encode(self.verifier.to_bytes());
-        let signature = STANDARD.encode(self.signer.sign(message.as_bytes()).to_bytes());
+        let verifying_key = STANDARD.encode(signing_key.verifying_key().to_bytes());
+        let signature = STANDARD.encode(signing_key.sign(message.as_bytes()).to_bytes());
 
         let subscribe_message = json!({
             "method": "SUBSCRIBE",
             "params": stream,
             "signature": [verifying_key, signature, timestamp.to_string(), window.to_string()],
         });
 
-        let ws_url = self.ws_url.as_deref().unwrap_or(BACKPACK_WS_URL);
+        let ws_url = self.ws_url.to_string();
```

Comment:
Nitpick: could just use `&str` here.

```suggestion
        let ws_url = self.ws_url.as_str();
```